src/awkward/_v2/highlevel.py

```diff
@@ -2085,726 +2085,721 @@ def __bool__(self):
+        """
+        Begins filling a record with an optional `name`; must be closed with
+        #end_record.
+
+        For example,
+
+            >>> builder = ak.ArrayBuilder()
+            >>> builder.begin_record("points")
+            >>> builder.field("x").real(1)
+            >>> builder.field("y").real(1.1)
+            >>> builder.end_record()
+            >>> builder.begin_record("points")
+            >>> builder.field("x").real(2)
+            >>> builder.field("y").real(2.2)
+            >>> builder.end_record()
+
+        produces
+
+            >>> ak.to_list(builder.snapshot())
+            [{"x": 1.0, "y": 1.1}, {"x": 2.0, "y": 2.2}]
+
+        with type
+
+            >>> ak.type(builder.snapshot())
+            2 * points["x": float64, "y": float64]
+
+        The record type is named `"points"` because its `"__record__"`
+        parameter is set to that value:
+
+            >>> builder.snapshot().layout.parameters
+            {'__record__': 'points'}
+
+        The `"__record__"` parameter can be used to add behavior to the records
+        in the array, as described in #ak.Array, #ak.Record, and #ak.behavior.
+        """
+        self._layout.beginrecord(name)
+
+    def field(self, key):
+        """
+        Args:
+            key (str): The field key to fill.
+
+        This method also returns the #ak.ArrayBuilder, so that it can be
+        chained with the value that fills the slot.
+
+        Prepares to fill a field; see #begin_record for an example.
+        """
+        self._layout.field(key)
+        return self
+
+    def end_record(self):
+        """
+        Ends a record.
+        """
+        self._layout.endrecord()
+
+    def append(self, obj, at=None):
+        """
+        Args:
+            obj: The object to append.
+            at (None or int): which value to select from `obj` if `obj` is
+                an #ak.Array.
+
+        Appends any type of object, which can be a shorthand for #null,
+        #boolean, #integer, #real, #bytestring, or #string, but also
+        an #ak.Array or #ak.Record to *reference* values from an existing
+        dataset, or any Python object to *convert* to Awkward Array.
+
+        If `obj` is an #ak.Array or #ak.Record, the output will be an
+        #ak.layout.IndexedArray64 (or #ak.layout.IndexedOptionArray64 if
+        there are any None values) that references the existing data. This
+        can be a more time and memory-efficient way to put old data in a
+        new structure, since it avoids copying and even walking over the
+        old data structure (matters more when the structures are large).
+
+        If `obj` is an arbitrary Python object, this is equivalent to
+        #ak.from_iter except that it fills an existing #ak.ArrayBuilder,
+        rather than creating a new one.
+
+        If `obj` is an #ak.Array and `at` is an int, this method fills the
+        ArrayBuilder with a reference to `obj[at]` instead of `obj`.
```

Comment:
This part of the documentation is no longer true; these paragraphs can just be dropped. If you want to mention that Python iterables (including dicts) are interpreted with #ak.from_iter, then that would be good.

For reference, here is the old, removed behavior:

```python
>>> import awkward as ak
>>> array = ak.Array([[1, 2, 3], [], [4, 5]])
>>> builder = ak.ArrayBuilder()
>>> builder.append(array)

>>> builder.snapshot()
<Array [[[1, 2, 3], [], [4, 5]]] type='1 * var * var * int64'>

>>> builder.snapshot().layout
<ListOffsetArray64>
    <offsets><Index64 i="[0 3]" offset="0" length="2" at="0x55ceeb42f9f0"/></offsets>
    <content><IndexedArray64>
        <index><Index64 i="[0 1 2]" offset="0" length="3" at="0x55ceeb847d60"/></index>
        <content><ListOffsetArray64>
            <offsets><Index64 i="[0 3 3 5]" offset="0" length="4" at="0x55ceeb117a30"/></offsets>
            <content><NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x55ceeb1143d0"/></content>
        </ListOffsetArray64></content>
    </IndexedArray64></content>
</ListOffsetArray64>
```

and here is the new behavior:

```python
>>> import awkward as ak
>>> array = ak.Array([[1, 2, 3], [], [4, 5]])
>>> builder = ak.ArrayBuilder()
>>> builder.append(array)

>>> builder.snapshot()
<Array [[[1, 2, 3], [], [4, 5]]] type='1 * var * var * int64'>

>>> builder.snapshot().layout
<ListOffsetArray64>
    <offsets><Index64 i="[0 3]" offset="0" length="2" at="0x56492e998cb0"/></offsets>
    <content><ListOffsetArray64>
        <offsets><Index64 i="[0 3 3 5]" offset="0" length="4" at="0x56492efd8e20"/></offsets>
        <content><NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x56492efd8f20"/></content>
    </ListOffsetArray64></content>
</ListOffsetArray64>
```

The difference is that there's no longer any link to the original array (through an IndexedArray); the `builder.append(array)` treats the `array` like any Python iterable and copies its data.